crypto/fipsmodule/bn/montgomery.c

```diff
@@ -418,6 +447,82 @@ static int bn_mod_mul_montgomery_fallback(BIGNUM *r, const BIGNUM *a,
   return ret;
 }
 
+
+#if defined(OPENSSL_BN_ASM_MONT)
+
+// Perform montgomery multiplication using s2n-bignum functions. The arguments
+// are equivalent to the arguments of bn_mul_mont.
+// montgomery_s2n_bignum_mul_mont works only if num is a multiple of 8. For
+// num = 32 or num = 16, this uses faster primitives in s2n-bignum.
+// Additionally, montgomery_s2n_bignum_mul_mont allocates arrays at a stack, and
+// large num leads to out of bounds accesses of the arrays.
+// montgomery_use_s2n_bignum(num) must be called in advance to check these
+// conditions.
+static void montgomery_s2n_bignum_mul_mont(BN_ULONG *rp, const BN_ULONG *ap,
+                                           const BN_ULONG *bp,
+                                           const BN_ULONG *np,
+                                           const BN_ULONG *n0, size_t num) {
+
+#if defined(BN_MONTGOMERY_USE_S2N_BIGNUM)
+
+  // t is a temporary buffer used by big-int multiplication.
+  // bignum_kmul_32_64 requires 96 words.
+  uint64_t t[96];
+  // mulres is the output buffer of big-int multiplication.
+  // If BN_MONTGOMERY_MAX_WORDS - 96 is larger than num*2, its low num*2
+  // elements are used.
+  // It is montgomery_use_s2n_bignum() that checks whether num*2 fits in the
+  // size of mulres array.
+  uint64_t mulres[BN_MONTGOMERY_MAX_WORDS - 96];
+
+  // Given m the prime number stored at np, m * w = -1 mod 2^64.
+  uint64_t w = n0[0];
+
+  if (num == 32) {
+    if (ap == bp)
+      bignum_ksqr_32_64(mulres, ap, t);
+    else
+      bignum_kmul_32_64(mulres, ap, bp, t);
+  } else if (num == 16) {
+    if (ap == bp)
+      bignum_ksqr_16_32(mulres, ap, t);
+    else
+      bignum_kmul_16_32(mulres, ap, bp, t);
+  } else {
+    if (ap == bp)
+      bignum_sqr(num * 2, mulres, num, ap);
+    else
+      bignum_mul(num * 2, mulres, num, ap, num, bp);
+  }
+
+  // Do montgomery reduction. We follow the definition of montgomery reduction
+  // which is:
+  // 1. Calculate (mulres + ((mulres mod R) * (-m^-1 mod R) mod R) * m) / R
+  //    using bignum_emontredc_8n, where R is 2^(64*num).
+  //    The calculated result is stored in [mulres+num ... mulres+2*num-1]. If
+  //    the result >= 2^(64*num), bignum_emontredc_8n returns 1.
+  // 2. Optionally subtract the result if the (result of step 1) >= m.
+  //    The comparison is true if either A or B holds:
+  //    A. The result of step 1 >= 2^(64*num), meaning that bignum_emontredc_8n
+  //       returned 1. Since m is less than 2^(64*num), (result of step 1) >= m holds.
+  //    B. The result of step 1 fits in 2^(64*num), and the result >= m.
+  uint64_t c;
+  c = bignum_emontredc_8n(num, mulres, np, w); // c: case A
+  c |= bignum_ge(num, mulres + num, num, np);  // c: case B
+  // Optionally subtract and store the result at rp
+  bignum_optsub(num, rp, mulres + num, c, np);
+
```

Comment:
is there a reason why the full montgomery reduction is not offered by s2n-bignum as a single function?

Reply:
There is the generic function bignum_montredc that does full Montgomery reduction (and indeed bignum_montmul doing full Montgomery multiplication), but that is not intended to be used for the performance-critical cases like this.

The argument for breaking this primitive out is that emontredc can also be a building-block for doing "almost-Montgomery" where you would do the subtraction only if you exceed the 2^64N size (that comparison is already the return value from the emontredc function). I think the whole modexp operation would be slightly faster maintaining this weaker normalization through the tower of multiplications (and just reducing strictly one time at the very end), but because we are trying to be plug-compatible at the level of BN_mul_montgomery, we do need to do the full reduction here every time, hence the additional bignum_ge.

Reply:
In addition to @jargh, I somewhat thought that the function invocations were straightforward since it was following the steps of montgomery reduction.. It was my thought though.

Reply:
Yes. `bignum_emontredc_8n` relies on the assumption that the input big integer is a multiple of 8 words for faster calculation. Specifically, it splits the input big integer into large chunks (4-words - which is 256 bits integers - that is a half of 8-words) and does fast multiplication on 256x256->512 bits, which internally uses Karatsuba multiplication. For Graviton 2, which has slow multiplier, using Karatsuba at this level brings significant performance benefit. :)

Reply:
There is a minor issue about the direction (even though I am aware that it is desirable to make things 'more correct') - it is that s2n-bignum does not (yet) support formal verification of linking of code. In this case linking is necessary because the three functions are defined in all different assembly files.
The reason why linking is unsupported is that (IIUC), linking requires the 'relocation' step which does not exist in s2n-bignum's assembly decoder yet. Adding this might be a fairly complicated process.

One possible workaround is to make a giant assembly file that copy-and-pastes the three assembly functions used in here. But this might require a really significant engineering effort. Proofs for these three functions are pretty large, and simply concatenating these proofs won't work in general due to... how it is written. Maybe this is why writing top-level, verified functions for elliptic curves is a big goal for this year as well (to avoid misinterpretation, I wanted to emphasize that verifying combined functions is non-trivial).

Of course we can put an unverified code that has the three function calls in s2n-bignum, but since s2n-bignum's policy is to always have a pair of code and proof, not sure whether this is a good direction too....

Reply:
I see. I’m ok with the compromise to get this code in as is with the expectation that s2n-bignum in the future will provide a single function call to do montgomery reduction (or even the full mul+redc).

Discussed with June offline, we agreed this is ok.